synapse_antispam/mjolnir/antispam.py

```diff
@@ -14,115 +14,145 @@
-                    return rule.action == RECOMMENDATION_BAN
-        return False
-
-    def is_server_banned(self, server_name):
-        for room_id in self.rooms_to_lists:
-            ban_list = self.rooms_to_lists[room_id]
-            for rule in ban_list.server_rules:
-                if rule.matches(server_name):
-                    return rule.action == RECOMMENDATION_BAN
-        return False
+        self._api = api
+        self._config = config
+
+        # Start a timer to rip the new rules off the server. We hit the rule server
+        # often to avoid missing rules - the response should be tens of milliseconds
+        # in duration, though network issues are possible - Twisted should automatically
+        # stack the calls for us.
+        # HACK: Private member access (_hs)
+        api._hs.get_clock().looping_call(self._update_rules, 5 * 1000)
+
+        # These are all arrays of compile()'d code from the ban list server.
+        # First match wins. See _update_rules() for more info.
+        self._code_spam_checks = []
+        self._code_invite_checks = []
+        self._code_profile_checks = []
+        self._code_create_room_checks = []
+        self._code_create_alias_checks = []
+        self._code_publish_room_checks = []
+
+    def _update_rules(self):
+        async def run():
+            try:
+                logger.info("Updating mjolnir rules...")
+
+                # HACK: Private member access (_hs)
+                resp = await self._api._hs.get_proxied_http_client().get_json(self._config['rules_url'])
+
+                # *** !! DANGER !! ***
+                # We're about to trust that the admin has secured their network appropriately and
+                # that the values returned by the configurable URL are safe to execute as real
+                # python code. We're using eval(), which should mean that it can only be expressions,
+                # however there is still risk of damages. We are knowingly processing the strings
+                # returned and assuming they won't try to take control of the homeserver, or worse.
+                # *** !! DANGER !! ***
+
+                self._code_spam_checks = self._compile_rules(resp['checks']['spam'])
+                self._code_invite_checks = self._compile_rules(resp['checks']['invites'])
+                self._code_profile_checks = self._compile_rules(resp['checks']['profiles'])
+                self._code_create_room_checks = self._compile_rules(resp['checks']['createRoom'])
+                self._code_create_alias_checks = self._compile_rules(resp['checks']['createAlias'])
+                self._code_publish_room_checks = self._compile_rules(resp['checks']['publishRoom'])
+
+                logger.info("Updated mjolnir rules")
+            except Exception as e:
+                logger.warning("Error updating mjolnir rules: %s", e)
+
+        return defer.ensureDeferred(run())
+
+    def _compile_rules(self, rules):
```

Comment:
It would help if the inputs and outputs of these methods were documented a bit. It seems that `rules` is a dictionary with a two keys:

* `search`: Python code as a string. It seems to be expected to be an expression that has access to some locally defined variables, which differ depending on the check being done.
* `pattern`: A string regular expression.

A couple of potential improvements:

* Use an attrs class instead of a dictionary for the returned lists (this should use less memory and allow for more efficient access to properties).
* Pre-compile the regular expressions using [`re.compile`](https://docs.python.org/3/library/re.html#re.compile) and then call `check["pattern"].search(search)`.